Sources/SwiftGRPC/Core/ClientNetworkMonitor.swift

```diff
@@ -0,0 +1,160 @@
+        let notificationName: Notification.Name
+        if #available(iOS 12.0, *) {
+            notificationName = .CTServiceRadioAccessTechnologyDidChange
+        } else {
+            notificationName = .CTRadioAccessTechnologyDidChange
+        }
+
+        NotificationCenter.default.addObserver(self, selector: #selector(self.cellularDidChange(_:)),
+                                               name: notificationName, object: nil)
+    }
+
+    @objc
+    private func cellularDidChange(_ notification: NSNotification) {
+        self.queue.async {
+            let newCellularName: String?
+            if #available(iOS 12.0, *) {
+                let cellularKey = notification.object as? String
+                newCellularName = cellularKey.flatMap { self.cellularInfo.serviceCurrentRadioAccessTechnology?[$0] }
+            } else {
+                newCellularName = notification.object as? String ?? self.cellularInfo.currentRadioAccessTechnology
+            }
+
+            if let newCellularName = newCellularName, self.cellularName != newCellularName {
+                self.cellularName = newCellularName
+                self.callback(State(lastChange: .cellularTechnology(technology: newCellularName),
+                                    isReachable: self.isReachable ?? false))
+            }
+        }
+    }
+
+    // MARK: - Reachability
+
+    private func startMonitoringReachability(_ reachability: SCNetworkReachability) {
+        let info = Unmanaged.passUnretained(self).toOpaque()
+        var context = SCNetworkReachabilityContext(version: 0, info: info, retain: nil,
+                                                   release: nil, copyDescription: nil)
+        let callback: SCNetworkReachabilityCallBack = { _, flags, info in
+            let observer = info.map { Unmanaged<ClientNetworkMonitor>.fromOpaque($0).takeUnretainedValue() }
+            observer?.reachabilityDidChange(with: flags)
+        }
+
+        SCNetworkReachabilitySetCallback(reachability, callback, &context)
+        SCNetworkReachabilityScheduleWithRunLoop(reachability, CFRunLoopGetMain(),
+                                                 CFRunLoopMode.commonModes.rawValue)
+        self.queue.async { [weak self] in
+            var flags = SCNetworkReachabilityFlags()
+            SCNetworkReachabilityGetFlags(reachability, &flags)
+            self?.reachabilityDidChange(with: flags)
+        }
+    }
+
+    private func reachabilityDidChange(with flags: SCNetworkReachabilityFlags) {
+        self.queue.async {
+            let isUsingWifi = !flags.contains(.isWWAN)
+            let isReachable = flags.contains(.reachable)
+
+            let notifyForWifi = self.isUsingWifi != nil && self.isUsingWifi != isUsingWifi
+            let notifyForReachable = self.isReachable != nil && self.isReachable != isReachable
+
+            self.isReachable = isReachable
+            self.isUsingWifi = isUsingWifi
+
+            if notifyForWifi {
```

Comment:
I think now this won't fire at all at launch. That's fine with me, but I guess then we could indeed fire twice if both `isReachable` and `isUsingWifi` have changed at the same time.

Reply:
Yea that makes sense. I'm leaning towards the latter since notifying on launch is probably not what the consumer wants if they're restarting channels based on this (since they'll likely end up restarting a brand new channel if we notify immediately)

Reply:
Updated to notify on either change